ext/opcache/jit/zend_jit_ir.c

```diff
@@ -16955,6 +17052,41 @@ static bool zend_jit_may_be_in_reg(const zend_op_array *op_array, zend_ssa *ssa,
 	return 1;
 }
 
+static ir_ref jit_frameless_observer(zend_jit_ctx *jit, const zend_op *opline, ir_ref op1_ref, uint32_t op1_info, ir_ref op2_ref, uint32_t op2_info, ir_ref op1_data_ref, uint32_t op1_data_info, ir_ref res_ref) {
+	// JIT: zend_observer_handler_is_unobserved(ZEND_OBSERVER_DATA(fbc))
+	ir_ref observer_handler;
+	zend_function *fbc = ZEND_FLF_FUNC(opline);
+	// Not need for runtime cache or generator checks here, we just need if_unobserved
+	ir_ref if_unobserved = jit_observer_fcall_is_unobserved_start(jit, fbc, &observer_handler, IR_UNUSED, IR_UNUSED).if_unobserved;
+
+	uint32_t call_num_args = ZEND_FLF_NUM_ARGS(opline->opcode);
+
+	// push call frame
+	ir_ref call_ref = ir_CALL_4(IR_ADDR, ir_CONST_FC_FUNC(zend_jit_observed_frameless_helper_push),
+		jit_FP(jit),
+		ir_CONST_U32(zend_vm_calc_used_stack(call_num_args, fbc)),
+		ir_CONST_ADDR(fbc),
+		ir_CONST_U32(call_num_args));
+
+	// push all args
+	switch (call_num_args) {
+		case 3: jit_ZVAL_COPY(jit, ZEND_ADDR_REF_ZVAL(ir_ADD_OFFSET(call_ref, EX_NUM_TO_VAR(2))), MAY_BE_ANY & ~MAY_BE_REF, ZEND_ADDR_REF_ZVAL(op1_data_ref), op1_data_info, 1); ZEND_FALLTHROUGH;
+		case 2: jit_ZVAL_COPY(jit, ZEND_ADDR_REF_ZVAL(ir_ADD_OFFSET(call_ref, EX_NUM_TO_VAR(1))), MAY_BE_ANY & ~MAY_BE_REF, ZEND_ADDR_REF_ZVAL(op2_ref), op2_info, 1); ZEND_FALLTHROUGH;
+		case 1: jit_ZVAL_COPY(jit, ZEND_ADDR_REF_ZVAL(ir_ADD_OFFSET(call_ref, EX_NUM_TO_VAR(0))), MAY_BE_ANY & ~MAY_BE_REF, ZEND_ADDR_REF_ZVAL(op1_ref), op1_info, 1);
+	}
+
+	// call and free args
+	ir_CALL_4(IR_VOID, ir_CONST_FC_FUNC(zend_jit_observed_frameless_helper_call),
+		call_ref,
+		ir_CONST_FC_FUNC(fbc->internal_function.handler),
+		observer_handler,
+		res_ref);
```

Comment:
Because I don't understand the JIT well enough. I don't know what state the JIT is in with respect to the VM.
Like, will the temporaries be actually set or something? Can I just fetch as if it were in VM? Like do EX_VAR() on any operand and the value is in correct state?
I thought it weren't unless we'd do deoptimization/exit from VM.
At least I don't feel confident doing it, but the way this currently is works very obviously to me.

Reply:
Okay, I managed to unify it and seems to work.